### PR TITLE
Release 1.2

### PR DESCRIPTION
**Pull request by StanislasGuinel**
This PR makes it possible to forecast after a new dataset instead of only the training dataset [sc-72262].
The training `TimeseriesPreparator`object is saved in the model folder: it now contains additional informations (prediction length and the timeseries identifiers values).
If a new dataset is provided in the 2. Forecast recipe (the historical_dataset), then the saved `TimeseriesPreparator` is retrieved from the model folder to check that all expected columns exist.
The new dataset is prepared using the same `TimeseriesPreparator` object as the training dataset, a GluonDataset is created and feed to the retrieved predictor.

Extra modifications:
- Custom models (AuroArima and SeasonalTrend) now store a dictionary of unique model per timeseries (values) by timeseries identifiers (key) instead of a list of models. It allows to use these models to predict only a specific timeseries identifier [sc-75735] instead of having to always predict all the timerseries ordered by their identifiers.
- Fix broken unit tests and add a new integration test
- Explicit error message if a timeseries identifiers is missing in the Dataset with future values of external features in the 2. Forecast recipe [sc-69777].

### CODE REVIEW

**Comment by StanislasGuinel** [comment]
[sc-69254] update gluonts

**Comment by StanislasGuinel** [comment]
[sc-72262] forecast after data not used during training

**Review by Marsobad — CHANGES_REQUESTED** [review]
Very good PR ! Congrats! 
I found some nitpicky naming comments and some edge cases. Also, we should add some unit tests to cover the most recent features
# Code logic
LGTM! Perhaps check the types in addition to the column name when loading the historical dataset. 
# Readability 
The code is easy to read and understand. The `GluonDataset` class could have a more specific name such as `DkuGluonDataset`. Also, naming some of the constants used could help make the code more accessible in the future
# Unit test coverage
Improve the unit test coverage by adding specific unit tests for the latest features (use historical dataset + one model for arima and seasonal trend models)
# Usage
The UI to select the input datasets does not help to pick the right datasets, but I know there's not much to do about it. Perhaps, in the documentation, we could add a visual presentation of each dataset that would include a timeline to help the users pick the right input datasets. Then add the link the description of the input dataset?

File: python-lib/timeseries_preparation/preparation.py [hunk 1]
```diff
@@ -67,9 +82,65 @@ def prepare_timeseries_dataframe(self, dataframe):
...
+    def check_schema_from_dataset(self, dataset_schema):
+        dataset_columns = [column["name"] for column in dataset_schema]
+        expected_columns = (
+            [self.time_column_name]
+            + (self.target_columns_names or [])
+            + (self.timeseries_identifiers_names or [])
+            + (self.external_features_columns_names or [])
+        )
+        if not set(expected_columns).issubset(set(dataset_columns)):
+            raise ValueError(f"Dataset of historical data must contain the following columns: {expected_columns}")
```
[the remark this answers is not in the document]
Comment (StanislasGuinel) [reply python-lib/timeseries_preparation/preparation.py:104]: I can move the check done in the TrainingSession class (it checks that the target and external features columns are numerical) to the TimeseriesPreparator class. This way we make sure it's numerical without ever enforcing the dss type (this plugin never looks at the dss types because it uses infer_with_pandas=True). 
The date column is already checked and for the identifiers column, it's even better than checking the types because we also check the values.

File: python-lib/gluonts_forecasts/gluon_dataset.py [hunk 2]
```diff
@@ -13,47 +12,53 @@ class GluonDataset:
         time_column_name (list)
         frequency (str): Pandas timeseries frequency (e.g. '3M')
         target_columns_names (list): List of column names to predict
-        timeseries_identifiers_names (list): Columns to identify multiple time series when data is in long format
-        external_features_columns_names (list): List of columns with dynamic real features over time
+        timeseries_identifiers_names (list, optional): Columns to identify multiple time series when data is in long format
+        external_features_columns_names (list, optional): List of columns with dynamic real features over time
```
[the remark this answers is not in the document]
Comment (StanislasGuinel) [reply python-lib/gluonts_forecasts/gluon_dataset.py:15]: done in b627de7

File: python-lib/gluonts_forecasts/gluon_dataset.py [hunk 3]
```diff
@@ -13,47 +12,53 @@ class GluonDataset:
         time_column_name (list)
```
[the remark this answers is not in the document]
Comment (StanislasGuinel) [reply python-lib/gluonts_forecasts/gluon_dataset.py:11]: done in b627de7

File: custom-recipes/timeseries-forecast-2-predict/recipe.py [hunk 4]
```diff
@@ -42,19 +44,42 @@ def run():
...
+
+        prepared_dataframe = timeseries_preparator.prepare_timeseries_dataframe(historical_dataset_dataframe)
+
+        gluon_dataset = GluonDataset(
+            time_column_name=timeseries_preparator.time_column_name,
+            frequency=timeseries_preparator.frequency,
+            target_columns_names=timeseries_preparator.target_columns_names,
+            timeseries_identifiers_names=timeseries_preparator.timeseries_identifiers_names,
+            external_features_columns_names=timeseries_preparator.external_features_columns_names,
+            min_length=2 * (timeseries_preparator.prediction_length if timeseries_preparator.prediction_length else 1),
```
[the remark this answers is not in the document]
Comment (StanislasGuinel) [reply custom-recipes/timeseries-forecast-2-predict/recipe.py]: done in b627de7

File: custom-recipes/timeseries-forecast-2-predict/recipe.py [hunk 5]
```diff
@@ -42,19 +44,42 @@ def run():
 
     predictor = model_selection.get_model_predictor()
 
-    gluon_train_dataset = model_selection.get_gluon_train_dataset()
+    if file_manager.historical_dataset:
+        # note: because dataframe is sorted in preparation according to identifiers and time column, order is maintained if same timeseries identifiers identifiers
```
[the remark this answers is not in the document]
Comment (StanislasGuinel) [reply custom-recipes/timeseries-forecast-2-predict/recipe.py]: note is now irrelevant, removed in b627de7

File: python-lib/dku_io_utils/checks_utils.py [hunk 6]
```diff
@@ -19,19 +19,23 @@ def external_features_future_dataset_schema_check(train_data_sample, external_fe
...
     """Check that an external features dataset with the right schema was provided if and only if external features were used during training
 
     Args:
-        gluon_train_dataset ([type]): [description]
-        external_features_future_dataset ([type]): [description]
+        gluon_train_dataset (ListDataset): Dataset used for training
+        external_features_future_dataset (dataiku.Dataset): Dataset of future values of external features
 
     Raises:
         ValueError:
```
[the remark this answers is not in the document]
Comment (StanislasGuinel) [reply python-lib/dku_io_utils/checks_utils.py:41]: done in b627de7

File: custom-recipes/timeseries-forecast-2-predict/recipe.py [hunk 7]
```diff
@@ -42,19 +44,42 @@ def run():
...
+            external_features_columns_names=timeseries_preparator.external_features_columns_names,
+            min_length=2 * (timeseries_preparator.prediction_length if timeseries_preparator.prediction_length else 1),
+        )
+
+        gluon_train_list_dataset = gluon_dataset.create_list_datasets(prepared_dataframe)[0]
+
+    else:
+        gluon_train_list_dataset = model_selection.get_gluon_train_list_dataset()
+
+    external_features = external_features_check(gluon_train_list_dataset, file_manager.external_features_future_dataset)
```
[the remark this answers is not in the document]
Comment (StanislasGuinel) [reply custom-recipes/timeseries-forecast-2-predict/recipe.py]: done in b627de7

File: python-lib/gluonts_forecasts/gluon_dataset.py [hunk 8]
```diff
@@ -13,47 +12,53 @@ class GluonDataset:
...
     def __init__(
         self,
-        dataframe,
         time_column_name,
         frequency,
         target_columns_names,
         timeseries_identifiers_names=None,
         external_features_columns_names=None,
-        min_length=None,
+        min_length=2,
```
[the remark this answers is not in the document]
Comment (StanislasGuinel) [reply python-lib/gluonts_forecasts/gluon_dataset.py]: done in b627de7

File: python-lib/dku_io_utils/model_selection.py [hunk 9]
```diff
@@ -83,12 +84,28 @@ def get_model_predictor(self):
...
+                self.folder, timeseries_preparator_path, ObjectType.JSON
+            )
+            timeseries_preparator = TimeseriesPreparator.deserialize(timeseries_preparator_serialized)
+
+        except ValueError as e:
+            raise ModelSelectionError(
+                f"Unable to retrieve training timeseries metadata at path '{timeseries_preparator_path}'"
+                f"from session '{self.session_name}'. "
+                + f"Please make sure that it exists in the Trained model folder. Full error: {e}"
+            )
```
[the remark this answers is not in the document]
Comment (StanislasGuinel) [reply python-lib/dku_io_utils/model_selection.py:107]: now also raise TypeError if some fields are bad or missing in the json, done in b627de7

File: custom-recipes/timeseries-forecast-2-predict/recipe.py [hunk 10]
```diff
@@ -7,13 +7,15 @@
 from dku_constants import RECIPE
 from gluonts_forecasts.utils import add_future_external_features
 from gluonts_forecasts.trained_model import TrainedModel
+from gluonts_forecasts.gluon_dataset import GluonDataset
```
[the remark this answers is not in the document]
Comment (StanislasGuinel) [reply custom-recipes/timeseries-forecast-2-predict/recipe.py]: done in a7f045e

File: tests/python/unit/test_autoarima.py [hunk 11]
```diff
@@ -58,7 +58,7 @@ def test_training_month_frequency(self):
         estimator = AutoARIMAEstimator(prediction_length=prediction_length, freq=frequency, max_D=3, season_length=4)
         predictor = estimator.train(gluon_dataset)
 
-        assert predictor.trained_models[0].seasonal_order[3] == 4  # seasonality is 4
+        assert predictor.trained_models[None].seasonal_order[3] == 4  # seasonality is 4
```
[the remark this answers is not in the document]
Comment (StanislasGuinel) [reply tests/python/unit/test_autoarima.py:61]: in 8081cad i test that the model predictor of autoarima and seasonal trend work with long format.

File: tests/python/unit/test_model.py [hunk 12]
```diff
@@ -166,7 +164,7 @@ def setup_class(self):
         )
 
         self.prediction_length = 2
-        gluon_list_datasets = gluon_dataset.create_list_datasets(cut_lengths=[self.prediction_length, 0])
+        gluon_list_datasets = gluon_dataset.create_list_datasets(df, cut_lengths=[self.prediction_length, 0])
```
[the remark this answers is not in the document]
Comment (StanislasGuinel) [reply tests/python/unit/test_model.py:167]: I'm not sure I could really make a unit test of that, it's more a feature for integration test as it involves multiple compoenents.
Actually, in test_trained_model.py, the TrainedModel object predicts using a gluon_dataset that wasn't used in training.

File: python-lib/timeseries_preparation/preparation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (StanislasGuinel) [reply python-lib/timeseries_preparation/preparation.py:104]: Done that in 3b2fee3

File: python-lib/dku_io_utils/model_selection.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Marsobad) [reply python-lib/dku_io_utils/model_selection.py:107]: Perfect!

File: python-lib/timeseries_preparation/preparation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Marsobad) [reply python-lib/timeseries_preparation/preparation.py:104]: alright!

**Review by Marsobad — APPROVED** [review]
LGTM!
